### PR TITLE
build: split webpack config as prod/dev

### PR DESCRIPTION
**Pull request by MaikoTan**

### CODE REVIEW

**Comment by MaikoTan** [comment]
Oh yeah, I almost forgot the config for gh-pages.
I'll add it later.

File: webpack/webpack.prod.cjs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+      },
+    ],
+  },
+};
+
+module.exports = mergeWithRules({
+  optimization: 'merge',
+  module: {
+    rules: {
+      test: 'match',
```
Comment (Trim21) [inline webpack/webpack.prod.cjs:51]: whats this？

File: webpack/webpack.prod.cjs [hunk 2]
```diff
@@ -0,0 +1,59 @@
+'use strict';
+
+const CssMinimizerPlugin = require('css-minimizer-webpack-plugin');
+const MiniCssExtractPlugin = require('mini-css-extract-plugin');
+const { mergeWithRules } = require('webpack-merge');
+
+const generateConfig = require(__dirname + '/webpack.config.cjs');
```
Comment (Trim21) [inline webpack/webpack.prod.cjs:7]: Why not just require with relative path?

Comment (MaikoTan) [reply webpack/webpack.prod.cjs:7]: People would run `npm` commands at the root of the repo, so the relative path IS `./webpack/webpack.config.cjs`. For consistency, I'll just concat the strings here.

File: webpack/webpack.prod.cjs (hunk 1 above)
Comment (MaikoTan) [reply webpack/webpack.prod.cjs:51]: See [webpack-merge#mergeWithRules](https://github.com/survivejs/webpack-merge#mergewithrules)

> `match` (`CustomizeRule.Match`) - Optional matcher that scopes merging behavior to a specific part based on similarity (think DOM or jQuery selectors)

File: webpack/webpack.prod.cjs (hunk 2 above)
Comment (Trim21) [reply webpack/webpack.prod.cjs:7]: a relative path is resolved from script file itself, not cwd... it's always `./webpack.config.cjs`

File: webpack/webpack.dev.cjs [hunk 3]
```diff
@@ -0,0 +1,46 @@
+'use strict';
+
+const { mergeWithRules } = require('webpack-merge');
+
+const generateConfig = require(__dirname + '/webpack.config.cjs');
+const constants = require(__dirname + '/constants.cjs');
```
Comment (Trim21) [inline webpack/webpack.dev.cjs:6]:
```suggestion
const generateConfig = require( './webpack.config.cjs');
const constants = require('./constants.cjs');
```

File: webpack/webpack.prod.cjs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+const constants = require(__dirname + '/constants.cjs');
```
Comment (Trim21) [inline webpack/webpack.prod.cjs:8]:
```suggestion
const generateConfig = require('./webpack.config.cjs');
const constants = require('./constants.cjs');
```

File: webpack/webpack.prod.cjs (hunk 2 above)
Comment (MaikoTan) [reply webpack/webpack.prod.cjs:7]: Ah, I tried and it worked.
I didn't remember why I add `__dirname` now, I'll change like that, Thanks.

File: webpack/webpack.dev.cjs [hunk 5]
```diff
@@ -39,8 +39,13 @@ module.exports = mergeWithRules({
       options: 'replace',
     },
   },
+  cache: 'replace',
 })(baseConfig, {
   mode: 'development',
   devtool: 'source-map',
+  cache: {
+    type: 'memory',
+    name: 'cactbot',
```
Comment (Trim21) [inline webpack/webpack.dev.cjs:48]:
```suggestion
-    name: 'cactbot',
```

`cache.name` only works with `type: 'filesystem'`

Comment (MaikoTan) [reply webpack/webpack.dev.cjs]: Wait?
In development mode, the `cache` property is set to `memory` by default, isn't it?
So I can delete this block entirely?

File: webpack/webpack.ghpages.cjs [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+const constants = require('./constants.cjs');
+
+// Add timerbar_test.html
+constants.cactbotModules['timerbarTest'] = 'ui/test/timerbar_test';
+constants.cactbotHtmlChunksMap['ui/test/timerbar_test.html'] = {
+  chunks: [
+    constants.cactbotModules.timerbarTest,
+  ],
+};
+
```
Comment (quisquous) [inline webpack/webpack.ghpages.cjs:15]: This is a little bit unfortunate to repeat, especially if we add more things like this in the future.  I think I'd be fine not shipping timerbar_test to gh-pages.

File: webpack/webpack.prod.cjs [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+const CssMinimizerPlugin = require('css-minimizer-webpack-plugin');
+const MiniCssExtractPlugin = require('mini-css-extract-plugin');
+const { mergeWithRules } = require('webpack-merge');
+
+const generateConfig = require('./webpack.config.cjs');
+const constants = require('./constants.cjs');
+
+const baseConfig = generateConfig(constants);
+
+const optimizationOverride = {
```
Comment (quisquous) [inline webpack/webpack.prod.cjs:12]: Why is the distributed release optimized but not gh-pages?

File: webpack/webpack.ghpages.cjs [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+const baseConfig = generateConfig(constants);
```
Comment (quisquous) [inline webpack/webpack.ghpages.cjs:16]: Hmm.  Is it possible to have ghpages be based on prod? Mostly I would like prod and ghpages to be pretty much identical, maybe minus the source map, and want changes to one to be reflected in the other.

File: webpack/webpack.ghpages.cjs (hunk 6 above)
Comment (MaikoTan) [reply webpack/webpack.ghpages.cjs:15]: Okay

File: webpack/webpack.prod.cjs (hunk 7 above)
Comment (MaikoTan) [reply webpack/webpack.prod.cjs:12]: Oh, yes. It should be optimized for gh-pages too.

File: webpack/webpack.ghpages.cjs (hunk 8 above)
Comment (MaikoTan) [reply webpack/webpack.ghpages.cjs:16]: Yes, I'll tweak it like that.

File: webpack/webpack.ghpages.cjs [hunk 9]
```diff
@@ -0,0 +1,9 @@
+'use strict';
+
+const { merge } = require('webpack-merge');
+
+const ProdConfig = require('./webpack.prod.cjs');
+
+module.exports = merge(ProdConfig, {
```
Comment (quisquous) [inline webpack/webpack.ghpages.cjs:7]:
```suggestion
const prodConfig = require('./webpack.prod.cjs');

module.exports = merge(prodConfig, {
```

File: webpack/webpack.config.cjs [hunk 10]
```diff
@@ -236,16 +107,11 @@ module.exports = function(env, argv) {
         {
           test: /\.ts$/,
           loader: 'ts-loader',
-          options: {
-            compilerOptions: dev ? { declaration: true, declarationMap: true } : {},
-          },
+          options: {},
```
Comment (quisquous) [inline webpack/webpack.config.cjs:110]: Is this needed?

**Review by quisquous — COMMENTED** [review]
Thanks for all the changes!

File: webpack/webpack.config.cjs (hunk 10 above)
Comment (MaikoTan) [reply webpack/webpack.config.cjs:110]: According to the doc from `webpack-merge`, it seems no need.
But I didn't test it.

Comment (MaikoTan) [reply webpack/webpack.config.cjs]: Tested, working.
So delete.

**Comment by quisquous** [comment]
image

This is me trying this patch via `npm start`.  Don't laugh at my crafting hotbars but I think CSS might not be loading?

**Comment by MaikoTan** [comment]
Oh, damn!
There lacks of a `style-loader`

**Comment by quisquous** [comment]
Images don't seem to work either (jobs, eureka).  Can you fix that up and then test a few overlays?

**Comment by MaikoTan** [comment]
> Images don't seem to work either (jobs, eureka).  Can you fix that up and then test a few overlays?

It is okay on my side. What problem did you meet?

**Comment by quisquous** [comment]
> > Images don't seem to work either (jobs, eureka).  Can you fix that up and then test a few overlays?
> 
> It is okay on my side. What problem did you meet?

Oh hmm, I can't reproduce it.  Sorry!

**Pull request merged**